Feat/multiple function access keys

petersalomonsen opened the pull request:
Preview URL: 

https://psalomobos.near.page/devhub.near/widget/app

Or see a demonstration video here:

https://youtu.be/CuwWv1zAKf4?si=lBUi9kXxRrhVjv-s

resolves #148 and https://github.com/near/neardevhub-widgets/issues/137

frol commented:
@petersalomonsen I don't understand what is the intended flow here. I don't see "don't ask me again" checkbox, when I tried to "like" on DevHub, there was a popup prompt that redirected me to the wallet (MNW) with some unparsable transaction confirmation (it tried to display it as a function call confirmation but there was no function call). Please, record a screencast as otherwise I don't even understand how you see it can work.

petersalomonsen commented:
You should see this:

<img width="1332" alt="image" src="https://github.com/NearSocial/VM/assets/9760441/97225b4a-e991-4566-af7a-c203c9836a59">

petersalomonsen commented:
Sorry @frol I must have missed something when I published the first version. Unfortunately what you experienced was that you were not redirected to the wallet for the `add_key` transaction, but instead just sent for the `add_like`. This is because of a missing `await` for the `add_key` transaction, and I've corrected this now.

Here's a demonstration video as requested: https://youtu.be/0oG9_-KEi0Q?si=BEeUYz5E7wQApWNm

petersalomonsen commented:
Thanks @frol , good to hear that we can move forward on this. I agree to all of your 3 points, and I've started working on the UX experience ( not using a browser prompt ).

As you can see in the screenshot below, from my local dev environment, there's a checkbox to toggle if you want to create a function access key ( we can consider naming this "Don't ask me again", as you propose ):

<img width="569" alt="image" src="https://github.com/NearSocial/VM/assets/9760441/f242fc4d-0baa-4b8c-9b16-49e7d0720826">

When you check that toggle, I'm now trying to include the function access key transaction, so that you don't loose the like or post you are about to confirm. So as you can see in the screenshot below, when checked, another transaction, the "addKey" is also showing.

<img width="543" alt="image" src="https://github.com/NearSocial/VM/assets/9760441/eb57b436-a677-487f-ae51-41d782bc4a55">

Not sure yet how to fix it so that the wallet says that "an application is requesting limited access to your account", but I'll try to figure that out.

Also totally agree that we should not reload the page, but right now I think this involves work on the DevHub code to handle transaction feedback. Everything we have right now is based on the fact that the page is reloading ( after a completed transaction ). I'll see what I can do there, but it might have to be a separate issue for DevHUB after this feature for creating function access keys is merged to the VM codebase.

Let me know what you think

petersalomonsen commented:
@frol @evgenykuzyakov I've done some UX tweaks now so that the adding of a new access key is part of the regular VM confirmation popup. Just as in the screenshots from the previous comment. Note that you only should tick the checkbox once, for the next calls you already have an access key stored, so you don't need to create it again. This might be confusing, so I will figure out how to make this more clear in the UI. "Don't ask again" might also be confusing since it could be misunderstood as you would not get the VM confirmation popup either.

kujtimprenkuSQA commented on src/lib/data/near.js:
```diff
@@ -82,6 +135,27 @@ async function functionCall(
   try {
     const wallet = await (await near.selector).wallet();
 
+    window.near = near;
```
I understand this is still in draft but would it be possible not to override `window.near` since it's used by Sender Wallet?

petersalomonsen replied:
yes, this was just for my own debugging, and will not be merged in with this PR. I will remove it!

petersalomonsen commented:
UX is improving bit by bit, so if you already have an access key stored, the popup will open with the checkbox checked, and also indicating which key that will be attempted for signing the transaction:

<img width="1042" alt="image" src="https://github.com/NearSocial/VM/assets/9760441/14e08d27-aaae-4d3f-b6a6-e6c62f229628">

frol commented:
> "Don't ask again" might also be confusing since it could be misunderstood as you would not get the VM confirmation popup either.

@petersalomonsen That is how SocialDB interactions are done - they allow “don’t ask again” for the specific component interacting with SocialDB. From UX perspective, I believe it makes sense. Obviously, you don’t want to allow any component to interact with a contract on behalf of the user just because there is already an access key, but it doesn’t make a lot of sense to confirm every like with the popup either. Take inspiration from Social.set implementation

petersalomonsen commented:
> > "Don't ask again" might also be confusing since it could be misunderstood as you would not get the VM confirmation popup either.
> 
> 
> 
> @petersalomonsen That is how SocialDB interactions are done - they allow “don’t ask again” for the specific component interacting with SocialDB. From UX perspective, I believe it makes sense. Obviously, you don’t want to allow any component to interact with a contract on behalf of the user just because there is already an access key, but it doesn’t make a lot of sense to confirm every like with the popup either. Take inspiration from Social.set implementation

Thanks @frol , that's a good point. I'll look into how Social.set is solved, and try avoiding the VM confirmation popup as well.

ailisp commented:
Looks great! I would also suggest to have this checked by default:
<img width="256" alt="image" src="https://github.com/NearSocial/VM/assets/13259400/879e7f8b-8ca5-46fe-88de-fcdf8a9932fe">

And I have a question when it's run out of function call key allowance, what will happen?

frol commented:
@petersalomonsen On a general note, please, put yourself in the shoes of a regular user, they don't know about "access key" and why they would want to create one.

petersalomonsen commented:
yes @frol . I'm moving forward in the direction you proposed with "Don't ask again", and not even showing the confirmation popup in the VM. @ailisp When the key runs out of allowance, we should detect that and pop up the confirmation dialogs again. It might be good to provide some optional detailed information for the user to explain what is happening behind the scenes.

petersalomonsen commented:
@frol @ailisp See the updated progress. Now it's using the regular wallet login, it's executing the "pending" transaction after login. After that there is no need to confirm when posting or liking. Just for now I've added a small popup ( just for information ), that shows that there is a transaction going on ( not sure how we want to solve this yet ).

Also see the new demo video of the flow: https://youtu.be/CuwWv1zAKf4?si=lBUi9kXxRrhVjv-s

The preview URL also contains the latest version.

petersalomonsen commented on src/lib/data/near.js:
```diff
@@ -111,33 +172,42 @@ async function accountState(near, accountId) {
 async function sendTransactions(near, functionCalls) {
+  if (functionCalls.length == 1) {
+    const { contractName, methodName, args, gas, deposit } = functionCalls[0];
+    //if (deposit == Big(0)) {
```
This check needs to be in place before merging, and also include the gas from the original transaction

petersalomonsen commented on src/lib/data/near.js:
```diff
@@ -111,33 +172,42 @@ async function accountState(near, accountId) {
           transactions.push({
             receiverId: contractName,
-            actions: [],
+            actions: actions ?? [],
```
actions is not used. should be removed

petersalomonsen commented on src/lib/data/near.js:
```diff
@@ -111,33 +172,42 @@ async function accountState(near, accountId) {
           currentTotalGas = newTotalGas;
         }
-        transactions[transactions.length - 1].actions.push(action);
+        if (!actions) {
```
there will not be actions here.

petersalomonsen commented on src/lib/data/near.js:
```diff
@@ -82,6 +126,23 @@ async function functionCall(
+          contractId: contractName,
+          methodName,
+          args,
+          //gas
```
need to include gas from the original transaction

petersalomonsen commented:
These are remaining, but could maybe be separate issues?

- Some indication on what's going on during sending the "Pending" transaction ( after login )
- Handling out of allowance
- Handling deleted function access keys ( that are not deleted in localstorage, but deleted on chain )

frol commented:
> Some indication on what's going on during sending the "Pending" transaction ( after login )

Well, given that Social.set also lacks any indication on Pending transaction, indeed, we can address it across the board in a separate issue.

> Handling out of allowance
> Handling deleted function access keys ( that are not deleted in localstorage, but deleted on chain )

`Social.set` also does not handle it gracefully either. Let's address these in separate issues, indeed. Please, create follow-up issues.

frol commented:
> After that there is no need to confirm when posting or liking.

I hope it only whitelists interaction with the contract only for the given BOS component. We don't want random components on BOS to be able to interact with devhub contract without user's acknowledgement. Again, see `Social.set` behavior.

<img width="1140" alt="image" src="https://github.com/NearSocial/VM/assets/304265/4f4d1166-75c3-4cde-bcc1-9339b59d399b">

(Also, let's use the same UI layout and message in both dialogs if possible)

> Just for now I've added a small popup ( just for information ), that shows that there is a transaction going on ( not sure how we want to solve this yet ).

Let's follow `Social.set`, it does not have any indication on BOS level, but exposes `onCommit` API (#105)

petersalomonsen commented:
Now I've fixed so that only interaction for the given BOS component is whitelisted, and you also only whitelist calling a specific method on the smart contract:

<img width="877" alt="image" src="https://github.com/NearSocial/VM/assets/9760441/b5cb0427-513a-42bf-8a4e-41f3f6cf9369">

Given all the changes, I do feel the need for setting up a testing framework for this now. Are there any automatic tests of the VM anywhere?

petersalomonsen commented:
Yes, the only thing I would like to see resolved before moving to ready for review is to remove deposit and set a lower gas amount in the devhub widgets as described here: https://github.com/near/neardevhub-widgets/issues/549 

Then I will be able to remove the overriding of deposit / gas that I had to do for demo purposes as you can see here:
https://github.com/NearSocial/VM/pull/156/files#diff-cc75698545c1506b569386ec69e64f50d829fc759a7fb175b256b88630782859R177

I will also go over the code a few more times and look for reusing of existing code.

petersalomonsen commented:
The purpose of the TransactionResultContext is to be able to show the results of the transaction without reloading the page , so that if you e.g. press "like" or add a post in DevHub, the page will update without reloading. The TransactionResultContext ensures that widgets are updated from the top of the hierarchy, since an application like DevHub has several nested widgets.

The purpose of the pending transaction is to also execute the transaction you intended to when prompted with "Don't ask again". In the case where you need to be redirected to the wallet for sigining into another contract, we also want to execute the transaction you intended to execute after that.

Is there any other way you would like to solve these?

frol commented:
> The purpose of the TransactionResultContext is to be able to show the results of the transaction without reloading the page 
> The purpose of the pending transaction is to also execute the transaction you intended to when prompted with "Don't ask again".

`Social.set` already solved both of these problems. Can you follow the pattern?

petersalomonsen commented:
> > The purpose of the TransactionResultContext is to be able to show the results of the transaction without reloading the page
> > The purpose of the pending transaction is to also execute the transaction you intended to when prompted with "Don't ask again".
> 
> `Social.set` already solved both of these problems. Can you follow the pattern?

All right it seems that invalidating the cache is enough, and then the widgets are updated after some time. I guess this is because I invalidate the cache for the entire devhub contract, and it takes some time to re-run. I'm not sure how to select which parts of the viewcall caches for a contract to be invalidated, since it's not obvious which data the contract will change from a transaction call. So I don't see any alternative than invalidating the whole cache for a contract. Do you see it different?

When it comes to the "pending" transaction, which is needed if you press "Like" in the devhub contract without having local access keys, and you need to be redirected to the wallet. You still want that "Like" to be executed right? I can't see that social.near has a solution for this. If you sign out of near.social, you can't commit transactions at all as far as I can see. Or should I be able to?

evgenykuzyakov commented:
So the challenge is to invalidate some cache, but it's unclear which once the transaction is confirmed. As for devhub calling `set` on `social.near` as a receipt. So the UI doesn't know what cache it has to invalidate.

In this case invalidating full `social.near` contract is also wrong, since many contract wouldn't touch `social.near` for their actions. So we should design a custom solution for devhub. One option is to introduce cache invalidation api, so the devhub UI can invalidate cache after a successful function call.

As for the pending transaction. Why not issue the transaction with the wallet redirect by requesting user to sign 2 transactions instead of one?

petersalomonsen commented:
> So the challenge is to invalidate some cache, but it's unclear which once the transaction is confirmed. As for devhub calling `set` on `social.near` as a receipt. So the UI doesn't know what cache it has to invalidate.
> In this case invalidating full `social.near` contract is also wrong, since many contract wouldn't touch `social.near` for their actions. So we should design a custom solution for devhub. One option is to introduce cache invalidation api, so the devhub UI can invalidate cache after a successful function call.

I'm not invalidating the `social.near` contract at all, as you can see [here](https://github.com/NearSocial/VM/pull/156/files#diff-78ae55da8050d0a4f8e77c9f862c6926d41f1f04dba33295c4d9e32d7f565196R184) I'm only invalidating the cache for the contract that was called ( for example the devhub contract ).

I think you suggestion about introducing a cache invalidation API is a good one. This will give more control to the UI on which cache to invalidate, but then I think a default behaviour of invalidating the cache for the entire contract is a good one too, in case the UI has not implemented anything. I suggest then maybe we merge what I have now, and we can introduce a cache invalidation API later?

> As for the pending transaction. Why not issue the transaction with the wallet redirect by requesting user to sign 2 transactions instead of one?

Requesting the user to sign 2 transactions was my original implementation. As you can see if you go some commits back, and from the screenshots in [this comment](https://github.com/NearSocial/VM/pull/156#issuecomment-1826836502), the user was asked to confirm two transactions. Also when redirected to the wallet the user was asked to confirm these two transactions, but as @frol pointed out, the user did not get a clear message that "an application is requesting limited access to your account". He then suggested to use the regular wallet sign-in flow, and execute the original transaction as a "pending transaction" after signing in to the new contract account. I think his suggestion makes sense, since sign-in experience is now more obvious to the user, compared to confirming a transaction about adding a function access key, which is not so clear in the wallet. Please discuss with @frol if you think we should rather revert to the original solution I had here, but in my opinion it's better now.

frol commented on src/lib/components/ConfirmTransactions.js:
```diff
@@ -12,78 +14,189 @@ ${json}
+  if (dontAskForConfirmation) {
+    const transaction = transactions[0];
+    return (<Modal size="xs" centered show={show} onHide={onHide}>
+      <Modal.Header>
+        <Modal.Title>Sending transaction</Modal.Title>
```
I feel it is quite distracting to show it as a modal. Ideally, I would use a [toast](https://getbootstrap.com/docs/5.3/components/toasts/).

petersalomonsen replied:
changed to using a toast

frol commented on src/lib/components/ConfirmTransactions.js:
```diff
@@ -12,78 +14,189 @@ ${json}
+        {Loading}
+      </Modal.Footer>
+    </Modal>);
+  } else if (dontAskForConfirmation == false) {
```
I guess this is enough:

```suggestion
  } else {
```

petersalomonsen replied:
`dontAskForConfirmation` is set by an async operation querying that the user is signed in and checking the widgets permission to call the contract without confirmation. The default state of this is `null`, while `false` confirms that we can show the confirmation popup.

frol replied:
@petersalomonsen This is hard to maintain it this way. I suggest you handle the `null` case first, and then handle only true/false (if/else) branches

frol commented on src/lib/components/ConfirmTransactions.js:
```diff
@@ -12,78 +14,189 @@ ${json}
+
+                if (!(await near.isSignedIntoContract(contractId))) {
+                  near.signInAndSetPendingTransaction(transactions[0]);
+                  return;
```
I feel that `setLoading(false)` and `onHide` needs to be called here to work fine with browser-extension wallets (e.g. Sender)

frol reviewed (changes requested):
> As for the pending transaction. Why not issue the transaction with the wallet redirect by requesting user to sign 2 transactions instead of one?

@evgenykuzyakov The UX of wallets do not make it clear enough when you try to add a limited access key as a transaction. "Login" is exactly the flow we want to trigger there. "Login and submit a transaction" might make sense to implement on the Wallets side, but at the same time it might be too custom. From UX perspective, I believe the current PR achieves the best possible one.

@petersalomonsen Please, test this implementation with Sender wallet and Meteor wallet that operate differently than MNW.

petersalomonsen commented on src/lib/data/near.js:
```diff
@@ -339,6 +404,16 @@ async function _initNear({
 
   _near.accountState = (accountId) => accountState(_near, accountId);
 
+  const pendingTransaction = JSON.parse(sessionStorage.getItem(PENDING_TRANSACTION_SESSION_STORAGE_KEY));
+  if (pendingTransaction) {
+    console.log('found pending transaction', pendingTransaction);
```
`sessionStorage` does only live in the current tab. If you open a new tab, then the pending transaction will not be known.

petersalomonsen commented on src/lib/components/ConfirmTransactions.js:
```diff
@@ -12,78 +14,189 @@ ${json}
+              e.preventDefault();
+              setLoading(true);
+              if (dontAskAgainChecked) {
+                const contractId = transactions[0].contractName;
+                const methodName = transactions[0].methodName;
```
As you see on line [139](https://github.com/NearSocial/VM/pull/156/files/22e8585d60da1b08eaa1889489bb48fa27f18341#diff-2816d3f924c1dbc749b117f468f07ebf91ea0e3fc8abe8dfa1bacd8e10162ea3R139) there is a check if the number of transactions is `1`. Otherwise, it will not offer the option to not ask again.

petersalomonsen commented on src/lib/data/near.js:
```diff
@@ -111,6 +166,13 @@ async function accountState(near, accountId) {
 
 async function sendTransactions(near, functionCalls) {
+  if (functionCalls.length == 1) {
+    const { contractName, methodName, args, gas, deposit } = functionCalls[0];
```
I can't imagine zero functionCalls here, but nothing is changed on how this is handled. Also here, there is a check if the number of functionCalls are 1. Otherwise the previous behaviour with confirmation through the wallet will apply

frol replied:
I was probably sleepy at that moment and somehow missed that

petersalomonsen commented on src/lib/data/near.js:
```diff
@@ -111,6 +166,13 @@ async function accountState(near, accountId) {
 }
 
 async function sendTransactions(near, functionCalls) {
+  if (functionCalls.length == 1) {
```
By "weirdo behavior" do you mean for example `"0xa"==10`  is `true`?

But of course, I will add the extra `=` to also ensure data type equality.

petersalomonsen commented on src/lib/data/near.js:
```diff
@@ -111,6 +166,13 @@ async function accountState(near, accountId) {
+  if (functionCalls.length == 1) {
+    const { contractName, methodName, args, gas, deposit } = functionCalls[0];
+
+    if (!deposit || deposit.toString() == '0') {
```
should be changed all places now

petersalomonsen commented:
> @petersalomonsen Please, test this implementation with Sender wallet and Meteor wallet that operate differently than MNW.

I've tested with Meteor wallet now, and since the wallet selector interfaces does not support signing into multiple contracts, and since Meteor wallet does not have a walletUrl, the `requestSignin`method of near-api-js will not work either. If not to modify the wallet selector here, the only option is to add a key through a transaction, and then both transactions might as well be combined. So for Meteor wallet it looks like below now. One transaction for generating the access key, and one for "like". Unfortunately this does not give a good login experience for this wallet. It just indicates a transaction to your own account, and if you expand the details, you see that it's adding the key.

<img width="944" alt="image" src="https://github.com/NearSocial/VM/assets/9760441/6a2e2036-6d15-4d94-b773-c635e16c4ea4">

as you can see in my [latest addition](https://github.com/NearSocial/VM/pull/156/files#diff-cc75698545c1506b569386ec69e64f50d829fc759a7fb175b256b88630782859R104), I check if the wallet has a URL, and in that case, I just use `walletConnection.requestSignin()`.. Otherwise adding a transaction for adding a key.

Not sure how to have the `requestSignin` experience also for MeteorWallet without modifying the wallet selector. I could do another PR for the wallet selector later to add support for this. What do you think @frol  @evgenykuzyakov ?

frol commented:
@petersalomonsen Thanks for the thorough testing.

> Sender does not support adding access keys in a transaction, and does not have a wallet url for near-api-js to redirect too, resulting in that the "Don't ask again" feature is not supported, and an error message is displayed instead.

Well, Sender has a different operation model and it breaks UX even for SocialDB interactions requiring explicit confirmation of every single transaction, so we cannot do anything here.

> I've tested with Meteor wallet now, and since the wallet selector interfaces does not support signing into multiple contracts, and since Meteor wallet does not have a walletUrl, the requestSigninmethod of near-api-js will not work either.

This is unexpected given that MeteorWallet is a browser wallet.

> Not sure how to have the requestSignin experience also for MeteorWallet without modifying the wallet selector. I could do another PR for the wallet selector later to add support for this.

I expected all the wallets (except Sender) to support the most standard login flow :-(

This rabbit hole is too deep. Please, create an issue on the Wallet Selector repo first and share it in the ["NEAR Wallet Builder Group"](https://nearbuilders.com/tg-wallet)

frol commented:
@evgenykuzyakov I think this current state already brings reasonable improvements to UX even if with some UX glitches on some of the wallets, so I suggest finishing the review and merging it as it is now and improving iteratively.

petersalomonsen commented:
> @petersalomonsen Thanks for the thorough testing.
> 
> > Sender does not support adding access keys in a transaction, and does not have a wallet url for near-api-js to redirect too, resulting in that the "Don't ask again" feature is not supported, and an error message is displayed instead.
> 
> Well, Sender has a different operation model and it breaks UX even for SocialDB interactions requiring explicit confirmation of every single transaction, so we cannot do anything here.
> 
> > I've tested with Meteor wallet now, and since the wallet selector interfaces does not support signing into multiple contracts, and since Meteor wallet does not have a walletUrl, the requestSigninmethod of near-api-js will not work either.
> 
> This is unexpected given that MeteorWallet is a browser wallet.
> 
> > Not sure how to have the requestSignin experience also for MeteorWallet without modifying the wallet selector. I could do another PR for the wallet selector later to add support for this.
> 
> I expected all the wallets (except Sender) to support the most standard login flow :-(
> 
> This rabbit hole is too deep. Please, create an issue on the Wallet Selector repo first and share it in the ["NEAR Wallet Builder Group"](https://nearbuilders.com/tg-wallet)

That is done now: https://github.com/near/wallet-selector/issues/1045

petersalomonsen commented:
Just discussed with @Megha-Dev-19 that [Near.call](https://docs.near.org/bos/api/near#nearcall) does not return anything today. This is already an issue with the wallets that does not redirect, but either open a separate window or extension for confirming the transaction. A widget does not have any way to know when a transaction is finished, or the result of it.

With this PR, the cache will be invalidated when the transaction is finished. This helps a lot, since it will trigger re-rendering. It would be even better though, if `Near.call` returned a promise with the result of the transaction.

Should I add a promise result to `Near.call` @evgenykuzyakov @frol ?

Megha-Dev-19 commented:
thanks @petersalomonsen for adding this, it helped me fixed other issues as well, now I can initiate an action after the user approves/reject the txn since the cache is invalidated.
@frol Would really appreciate if we can get this merged asap 🙏

frol commented:
> It would be even better though, if Near.call returned a promise with the result of the transaction.
> Should I add a promise result to Near.call @evgenykuzyakov @frol ?

I think we should mimic Social.set behavior with the `onCommit` and `onCancel` callbacks as you suggested in #105. I would make it a separate PR.

frol commented on src/lib/components/ConfirmTransactions.js:
```diff
@@ -12,78 +17,201 @@ ${json}
+        </Toast>
+      </ToastContainer>
+    );
+  } else if (dontAskForConfirmation === false) {
```
As I mentioned before (https://github.com/NearSocial/VM/pull/156#discussion_r1425768505), please, refactor it so it is easier to maintain this code in the future as to me `if (x) {} else if (x === false) {} else {}` is really hard to reason about and I imagine future developers will make wrong assumptions about this code. Replace it with:

```ts
if (x === null) {
  return ...;
}
if (x) {
} else {
}
```

petersalomonsen replied:
Good point. I've moved the check for `null` to the top now.

I would also suggest introducing a test framework to avoid future regressions, but maybe we want a separate PR for that?

frol replied:
yeah, regression testing should be implemented separately

petersalomonsen replied:
From another point of view. If there will still be some time before this will be merged, I could add tests so that we have "proof" of the scenarios covered here. It's always reassuring to deliver a PR with tests to prove that what you deliver works as it should. Let me know what you think.

frol commented on src/lib/components/ConfirmTransactions.js:
```diff
@@ -12,78 +17,201 @@ ${json}
+
+        const widgetContractPermission = await getWidgetContractPermission(widgetSrc, contractId);
+
+        const dontAskForConfirmation = (isSignedIntoContract && widgetContractPermission && widgetContractPermission[transactions[0].methodName])
+          ? true : false;
```
Function calls with non-zero deposits cannot be signed with a limited access key (i.e. "Don't ask again" is not even applicable), and I feel this code does not handle this situation properly.

petersalomonsen replied:
I have added a check also for non-zero deposits now, so that the user is not offered "Don't ask again". Also if a previously approved similar transaction ( same contract, same method name ), now has a deposit, then it will require confirmation.

petersalomonsen commented:
> > It would be even better though, if Near.call returned a promise with the result of the transaction.
> > Should I add a promise result to Near.call @evgenykuzyakov @frol ?
> 
> I think we should mimic Social.set behavior with the `onCommit` and `onCancel` callbacks as you suggested in #105. I would make it a separate PR.

Thanks for reminding me, I had completely forgotten that I wrote that. There is already a PR for that also.

frol reviewed (approved):
Looks good to me.

@evgenykuzyakov  please, review once again

evgenykuzyakov commented on src/lib/data/near.js:
```diff
@@ -71,6 +73,98 @@ const apiCall = async (config, methodName, args, blockId, fallback) => {
+  );
+}
+
+async function createWalletConnectionForContract(near, contractId) {
+  const keyStore = getKeyStoreForContract(contractId);
```
We should probably check somewhere that the contractId doesn't match `social.near` (for the current network) to avoid some random issues.

petersalomonsen replied:
There is already such a check on `functionCall` ( see https://github.com/NearSocial/VM/pull/156/files#diff-cc75698545c1506b569386ec69e64f50d829fc759a7fb175b256b88630782859R179 )

evgenykuzyakov replied:
Do we check for this in the `ConfirmTransaction`?

petersalomonsen replied:
No. And I believe if call a function on `social.near`, you might be offered `don't ask again`. This could be solved by also adding this check to `eligibleForDontAskAgain`.

Do we want to **not** offer the `don't ask again` feature for calls to `near.social`?

evgenykuzyakov replied:
Yeah, let's not give this option for `social.near`.

evgenykuzyakov commented on src/lib/components/Widget.js:
```diff
@@ -4,6 +4,7 @@ import React, {
   useEffect,
   useLayoutEffect,
   useState,
+  createContext
```
This is not needed, right?

evgenykuzyakov reviewed (commented):
Looks good. Minor comments

petersalomonsen commented:
In the latest commit, there's also a try/catch when sending the pending transaction, so that if the transaction fails, the page will still be loaded ( without this, the page would just be blank, and reloading would not help ).

evgenykuzyakov commented:
Please rebase to `dev` branch

petersalomonsen commented:
> Please rebase to `dev` branch

Rebased to `upstream/dev`, and also a quick re-test. Also updated preview environment with the latest rebased: https://psalomobos.near.page/devhub.near/widget/app?page=post&id=2688

evgenykuzyakov commented:
Let's remove don't ask again for `social.near`

petersalomonsen commented:
> Let's remove don't ask again for `social.near`

OK this is now implemented in the [latest push](https://github.com/NearSocial/VM/compare/24d35214693822cce0b17cd2bb02a834e3fe8eeb..895a29467422f82076b42f212b7f92e55231006d), and can be tested [here](https://psalomobos.near.page/petersalomonsen.near/widget/callsocialneartestdontaskagain), where you see that calling `social.near` does not present the "Don't ask again" option, while calling another contract does.

<img width="390" alt="image" src="https://github.com/NearSocial/VM/assets/9760441/80317114-65c9-4c45-999a-d47b9ebcfb0b">
<img width="414" alt="image" src="https://github.com/NearSocial/VM/assets/9760441/3a982bcb-33c5-47aa-acff-d4e22fcd13e7">
